pkg/job/associator.go

```diff
@@ -2,66 +2,168 @@ package job
+//
+// 1. First, it will only consider the dimensions from the metric that were seen by the associator. This allows matching
+// a metric that targets a specific resource, but is scoped over a property of it. For example, GlobalAccelerator's
+// `NewFlowCount` metric scoped just over the `tcp` connections started in an accelerator. This will contain the
+// `Accelerator` and `TransportProtocol` dimensions.
+// 2. For that given set of dimensions, it will check if the associator contains any resources. If there's none, the metric
+// will still be used.
+// 3a. If that set of dimensions is known by the associator, and there is a resource that matches all, it will be returned.
+// 3b. If not, nil will be returned, and the metric should be omitted.
+func (a *associator) associateMetricsToResources(cwMetric *cloudwatch.Metric) (*model.TaggedResource, bool) {
+	matches := []match{}
+	metricDimensions := make(stringSet)
+	for _, dim := range cwMetric.Dimensions {
+		metricDimensions.add(*dim.Name)
+	}
+	intersect := a.seenDimensions.intersect(metricDimensions)
+	for _, dim := range cwMetric.Dimensions {
+		// only consider dimensions that are both in the ones seen by the associator, and the current metric. This gives
+		// the algorithm the best effort matching
+		if intersect.contains(*dim.Name) {
+			matches = append(matches, match{
+				name:  *dim.Name,
+				value: *dim.Value,
+			})
+		}
+	}
+	encodedDimensions, encodedValues := encodeMatches(matches)
+	// if the dimension set of which we are looking a resource doesn't exist, return nil but avoid skipping the metric we
+	// are matching. This is the default logic, and it will associate with a generic resource
+	if _, ok := a.associations[encodedDimensions]; !ok {
+		return nil, false
+	}
+
+	// the dimension set exists in the associator, so there needs to be a match in order for the metrics to be used
+	if matchedResource, ok := a.associations[encodedDimensions][encodedValues]; ok {
+		return matchedResource, false
 	}
-	return dimensionsFilter
+	return nil, true
 }
 
-// associateMetricsToResources finds for a cloudwatch.Metrics, the resource that matches the better. If no match is found,
-// nil is returned. Also, there's some conditions in which the metric shouldn't be considered, and that is dictated by the
-// skip return value.
-func (asoc metricsToResourceAssociator) associateMetricsToResources(cwMetric *cloudwatch.Metric) (r *model.TaggedResource, skip bool) {
-	alreadyFound := false
-	for _, dimension := range cwMetric.Dimensions {
-		if dimensionFilterValues, ok := asoc[*dimension.Name]; ok {
-			if d, ok := dimensionFilterValues[*dimension.Value]; !ok {
-				if !alreadyFound {
-					skip = true
-				}
-				break
-			} else {
-				alreadyFound = true
-				r = d
-			}
+const presentByte = byte('#')
+
+// stringSet is a simple set implementation, with string values, that allows intersection operations.
+type stringSet map[string]byte
```

Comment:
I think this is functionally the same as `map[string]struct{}`, right? I think empty struct is preferred due to no allocations. I think this one has very minimal allocations but threw me off for a little until I realized the value wasn't used.

Reply:
yep, can definitely use `map[string]struct{}`!

But if you refer `stringSet` having the same functionality as the map it doesn't, since it allows one to intersect "sets"

Reply:
Yep, I was only referring to the type implementation. No issue at all having the dedicated type with the intersect functionality.